src/scripts/train/learn2code/love2code-3B.py

```diff
@@ -0,0 +1,301 @@
+from olmo_core.distributed.parallel import DataParallelType
+from olmo_core.internal.experiment import CommonComponents, main, build_common_components, SubCmd, ExperimentConfig
+from olmo_core.internal.common import get_work_dir, get_root_dir
+from olmo_core.train.common import Duration
+from olmo_core.nn.transformer import TransformerConfig, TransformerDataParallelConfig
+from olmo_core.optim import AdamWConfig, OptimGroupOverride, OptimConfig
+from olmo_core.data import (
+    DataMixBase,
+    NumpyDataLoaderConfig,
+    NumpyDatasetConfig,
+    TokenizerConfig,
+    TokenizerName,
+)
+from olmo_core.train import TrainerConfig
+from olmo_core.train.callbacks import CheckpointerCallback, CometCallback, WandBCallback
+from typing import Callable, Dict, List, Tuple, Optional, cast
+import sys
+
+SEQUENCE_LENGTH = 2048 
+CHINCHILLA_5X_DURATION = Duration.tokens(36072678400 * 20 * 5)
+
+# =========================================================
+# =                 "COMMON COMPONENTS" STUFF             =
+# =========================================================
+
+
+class Love2CodeDataMix(DataMixBase):
+    """ Defines love2code mix. To create a new mix, make a new file in this folder
+        and its name (without the '.txt' extension) below.
+
+    NOTE FOR DG/PW: I based this off of the OLMo2-32B-anneal.py AnnealingDataMix class
+    """
+
+    love2code_mix = 'love2code_data_XB.txt'
+    def build(self, base_dir: str, tokenizer: str) -> Tuple[List[str], List[str]]:
+        if not base_dir.endswith("/"):
+            base_dir = base_dir + "/"
+
+        assert tokenizer == TokenizerName.dolma2
+        paths = []
+        labels = []
+
+        with open(f"src/scripts/train/learn2code/{self}.txt") as f:
+            for line in f:
+                line = line.strip()
+                if not line or line.startswith("#"):
+                    continue
+                paths.append(f"{base_dir}{line}")
+                labels.append(line.split("/")[1])
+
+        return paths, labels    
+
+
+def build_love2code_common(
+    script: str,
+    cmd: SubCmd,
+    run_name: str,
+    overrides: List[str],
+    *,
+    global_batch_size: int
+) -> CommonComponents:
+    """ Note for DG/PW: 
+
+    I took the original internal.experiment.build_common_components and ran it
+    and then built the dataset config and hotswapped it. Maybe not canonical, but seems easiest?
+    """
+    og_common = build_common_components(script, cmd, run_name, cluster, overrides, global_batch_size=global_batch_size)
+    tokenizer_config = og_common.tokenizer
+    root_dir = get_root_dir(cluster)
+
+    dataset = NumpyDatasetConfig.from_data_mix(
+        Love2CodeDataMix.love2code_mix,
+        tokenizer=tokenizer,
+        mix_base_dir=root_dir,
+        sequence_length=SEQUENCE_LENGTH,
+        work_dir=get_work_dir(root_dir)
+        )
+    data_loader_config = NumpyDataLoaderConfig(
+        global_batch_size=global_batch_size, seed=34521, num_workers=4)
```

Comment:
Guessing you need more data loading workers. A small model will run fast, so the data needs to come in faster. I'd start with 8 or 16.